cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -0,0 +1,258 @@
+func (gc *GceCache) RegisterMig(mig Mig) bool {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	for i := range gc.migs {
+		if oldMig := gc.migs[i].Config; oldMig.GceRef() == mig.GceRef() {
+			if !reflect.DeepEqual(oldMig, mig) {
+				gc.migs[i].Config = mig
+				glog.V(4).Infof("Updated Mig %s", mig.GceRef().String())
+				return true
+			}
+			return false
+		}
+	}
+
+	glog.V(1).Infof("Registering %s", mig.GceRef().String())
+	gc.migs = append(gc.migs, &MigInformation{
+		Config: mig,
+	})
+	return true
+}
+
+// UnregisterMig returns true if the node group has been removed, and false if it was alredy missing from cache.
+func (gc *GceCache) UnregisterMig(toBeRemoved Mig) bool {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	newMigs := make([]*MigInformation, 0, len(gc.migs))
+	found := false
+	for _, mig := range gc.migs {
+		if mig.Config.GceRef() == toBeRemoved.GceRef() {
+			glog.V(1).Infof("Unregistered Mig %s", toBeRemoved.GceRef().String())
+			found = true
+		} else {
+			newMigs = append(newMigs, mig)
+		}
+	}
+	gc.migs = newMigs
+	return found
+}
+
+// GetMigs returns a copy of migs list.
+func (gc *GceCache) GetMigs() []*MigInformation {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	migs := make([]*MigInformation, 0, len(gc.migs))
+	for _, mig := range gc.migs {
+		migs = append(migs, &MigInformation{
+			Basename: mig.Basename,
+			Config:   mig.Config,
+		})
+	}
+	return migs
+}
+
+func (gc *GceCache) updateMigBasename(ref GceRef, basename string) {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	for _, mig := range gc.migs {
+		if mig.Config.GceRef() == ref {
+			mig.Basename = basename
+		}
+	}
+}
+
+// Methods locking on cacheMutex.
+
+// GetMigForInstance returns MigConfig of the given Instance from cache.
+// Attempts to regenerate cache if there is a Mig with matching prefix in migs list.
+// TODO(aleksandra-malinowska): reconsider failing when there's a Mig with
+// matching prefix, but instance doesn't belong to it.
+func (gc *GceCache) GetMigForInstance(instance *GceRef) (Mig, error) {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	if mig, found := gc.migCache[*instance]; found {
+		return mig, nil
+	}
+
+	for _, mig := range gc.GetMigs() {
+		if mig.Config.GceRef().Project == instance.Project &&
+			mig.Config.GceRef().Zone == instance.Zone &&
+			strings.HasPrefix(instance.Name, mig.Basename) {
+			if err := gc.regenerateCache(); err != nil {
+				return nil, fmt.Errorf("Error while looking for MIG for instance %+v, error: %v", *instance, err)
+			}
+			if mig, found := gc.migCache[*instance]; found {
+				return mig, nil
+			}
+			return nil, fmt.Errorf("Instance %+v does not belong to any configured MIG", *instance)
+		}
+	}
+	// Instance doesn't belong to any configured mig.
+	return nil, nil
+}
+
+// RegenerateCacheWithLock triggers cache regeneration under lock.
+func (gc *GceCache) RegenerateCacheWithLock() error {
```

Comment:
I don't like this method name. It's the only way to regenerate cache that is externally available and the caller shouldn't need to know about internal locking. It should be called RegenerateCache or even better RegenerateInstanceCache.

Reply:
Done. Also renamed migCache to instancesCache for consistency.